File: android/app/src/main/java/app/spotter/dev/ContactSheet.java
package app.spotter.dev;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.RectF;
import android.media.MediaMetadataRetriever;
import android.os.Build;

import java.io.ByteArrayOutputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * The phone cuts the frames — the Android half.
 *
 * One real difference from iOS, and it is in Android's favour:
 * MediaMetadataRetriever.setDataSource(url, headers) re-sends those headers on
 * every HTTP range request it makes (AOSP MediaHTTPConnection), so the MP4 never
 * has to be downloaded at all. The retriever reads the ranges it needs around
 * each keyframe straight off TikTok's CDN with the Cookie and Referer that make
 * the CDN answer, and nothing lands on disk to be forgotten about.
 *
 * The other difference is a cost: the retriever will not say WHICH frame it gave
 * you. OPTION_CLOSEST_SYNC is the fast path (OPTION_CLOSEST decodes forward from
 * the previous keyframe and is far too slow for a save), but it snaps to the
 * nearest sync sample without reporting where that was, so the label says the
 * time that was asked for, within one keyframe interval of the picture — and two
 * neighbouring requests can land on the same keyframe and return the same
 * picture twice. Hence the sameAs() check: a duplicate cell is wasted attention,
 * so the second copy is dropped and its time with it.
 *
 * Every number comes from SheetSpec, which is asserted against
 * native/sheet-spec.json — and therefore against the iOS build — by
 * tools/android/sheet-check.mjs.
 */
public final class ContactSheet {
    private ContactSheet() {}

    /** Below this a sheet says less than nothing; the save goes without frames. */
    public static final int MIN_USABLE_FRAMES = 4;

    /** One sheet: the JPEG, and the times of the cells on it, in cell order. */
    public static final class Page {
        public final byte[] jpeg;
        public final double[] times;
        public final int cols, rows;
        Page(byte[] jpeg, double[] times, int cols, int rows) {
            this.jpeg = jpeg; this.times = times; this.cols = cols; this.rows = rows;
        }
    }

    public static final class Result {
        public final List<Page> pages;
        public final int cellW, cellH, requested;
        public final double durationS;
        Result(List<Page> pages, int cellW, int cellH, int requested, double durationS) {
            this.pages = pages; this.cellW = cellW; this.cellH = cellH;
            this.requested = requested; this.durationS = durationS;
        }
    }

    /**
     * @param source  a remote MP4 URL, or an absolute path to a local file
     * @param headers Cookie / Referer / User-Agent for the remote case; null locally
     * @param deadline System.currentTimeMillis() past which whatever has been cut is the sheet
     */
    public static Result build(String source, Map<String, String> headers, long deadline) throws Exception {
        MediaMetadataRetriever retriever = new MediaMetadataRetriever();
        List<Bitmap> chunk = new ArrayList<>();
        List<Double> chunkTimes = new ArrayList<>();
        List<Page> pages = new ArrayList<>();
        try {
            if (headers != null && !headers.isEmpty()) retriever.setDataSource(source, headers);
            else retriever.setDataSource(source);

            double duration = number(retriever, MediaMetadataRetriever.METADATA_KEY_DURATION) / 1000.0;
            double width = number(retriever, MediaMetadataRetriever.METADATA_KEY_VIDEO_WIDTH);
            double height = number(retriever, MediaMetadataRetriever.METADATA_KEY_VIDEO_HEIGHT);
            int rotation = (int) number(retriever, MediaMetadataRetriever.METADATA_KEY_VIDEO_ROTATION);
            // A portrait phone video is often stored landscape with a 90 degree
            // rotation flag; the frames come back upright, so the box has to be
            // chosen from the upright dimensions rather than the stored ones.
            if (rotation == 90 || rotation == 270) { double swap = width; width = height; height = swap; }
            if (duration <= 0) throw new IllegalStateException("no duration");

            int[] cell = SheetSpec.cell(width, height);
            double[] wanted = SheetSpec.times(duration);
            Bitmap previous = null;
            int kept = 0, decoded = 0;

            // Twelve cells make a sheet, and the sheet is drawn and its frames
            // released before the next twelve are asked for: peak memory is one
            // sheet's worth of bitmaps, not three.
            for (double t : wanted) {
                if (System.currentTimeMillis() >= deadline || pages.size() >= SheetSpec.MAX_SHEETS) break;
                Bitmap frame = frameAt(retriever, (long) (t * 1_000_000L), cell[0], cell[1]);
                if (frame == null) continue;
                decoded++;
                if (previous != null && frame.sameAs(previous)) { frame.recycle(); continue; }
                chunk.add(frame);
                chunkTimes.add(t);
                // Only a reference for the next sameAs() test — the bitmap itself
                // belongs to the chunk until flush() has drawn and released it.
                previous = frame;
                kept++;
                if (chunk.size() == SheetSpec.CELLS_PER_SHEET) {
                    pages.add(flush(chunk, chunkTimes, cell));
                    previous = null;
                }
            }
            if (!chunk.isEmpty() && pages.size() < SheetSpec.MAX_SHEETS) {
                pages.add(flush(chunk, chunkTimes, cell));
                previous = null;
            }
            // An incomplete decode is not a complete overview; use the existing fallback.
            if (decoded != wanted.length || kept < MIN_USABLE_FRAMES || pages.isEmpty()) throw new IllegalStateException("too few frames");
            return new Result(pages, cell[0], cell[1], wanted.length, duration);
        } finally {
            for (Bitmap b : chunk) if (!b.isRecycled()) b.recycle();
            try { retriever.release(); } catch (Exception ignored) {}
        }
    }

    /** Draw the frames held so far into a sheet and let go of them. */
    private static Page flush(List<Bitmap> chunk, List<Double> chunkTimes, int[] cell) {
        double[] times = new double[chunkTimes.size()];
        for (int i = 0; i < times.length; i++) times[i] = chunkTimes.get(i);
        byte[] jpeg = draw(chunk, times, cell);
        for (Bitmap b : chunk) if (!b.isRecycled()) b.recycle();
        chunk.clear();
        chunkTimes.clear();
        return new Page(jpeg, times, Math.min(SheetSpec.COLS, times.length), SheetSpec.rows(times.length));
    }

    private static byte[] draw(List<Bitmap> frames, double[] times, int[] cell) {
        int[] size = SheetSpec.canvas(times.length, cell);
        Bitmap sheet = Bitmap.createBitmap(size[0], size[1], Bitmap.Config.ARGB_8888);
        try {
            Canvas canvas = new Canvas(sheet);
            // Black under everything: the gutters, the bars beside an aspect-fitted
            // frame, and any cell the last row does not reach.
            canvas.drawColor(Color.BLACK);

            Paint image = new Paint(Paint.FILTER_BITMAP_FLAG);
            Paint pill = new Paint(Paint.ANTI_ALIAS_FLAG);
            pill.setColor(Color.argb((int) Math.round(SheetSpec.LABEL_PILL_ALPHA * 255), 0, 0, 0));
            Paint text = new Paint(Paint.ANTI_ALIAS_FLAG);
            text.setColor(Color.WHITE);
            text.setTextSize((float) SheetSpec.LABEL_FONT_SIZE);
            text.setFakeBoldText(true);
            float ascent = text.ascent(), descent = text.descent();
            float lineHeight = descent - ascent;

            for (int i = 0; i < frames.size(); i++) {
                Bitmap frame = frames.get(i);
                // The slot is the cell inset by half a gutter on every side, so two
                // neighbouring frames are two black pixels apart and the sheet is
                // still exactly cols*cell_w by rows*cell_h.
                int[] slot = SheetSpec.frame(i, cell);
                // Fit, never fill: a crop is exactly where the second kettlebell at
                // the edge of the frame would have gone missing.
                float scale = Math.min(slot[2] / (float) frame.getWidth(), slot[3] / (float) frame.getHeight());
                float w = frame.getWidth() * scale, h = frame.getHeight() * scale;
                float left = slot[0] + (slot[2] - w) / 2f, top = slot[1] + (slot[3] - h) / 2f;
                canvas.drawBitmap(frame, new Rect(0, 0, frame.getWidth(), frame.getHeight()),
                        new RectF(left, top, left + w, top + h), image);

                // The decoder cannot report actual PTS; visibly mark this as approximate.
                String label = "~" + SheetSpec.label(times[i]);
                float pillW = text.measureText(label) + (float) SheetSpec.LABEL_PAD_X * 2;
                float pillH = lineHeight + (float) SheetSpec.LABEL_PAD_Y * 2;
                float pillX = slot[0] + (float) SheetSpec.LABEL_INSET;
                float pillY = slot[1] + slot[3] - (float) SheetSpec.LABEL_INSET - pillH;
                canvas.drawRoundRect(new RectF(pillX, pillY, pillX + pillW, pillY + pillH),
                        (float) SheetSpec.LABEL_PILL_RADIUS, (float) SheetSpec.LABEL_PILL_RADIUS, pill);
                canvas.drawText(label, pillX + (float) SheetSpec.LABEL_PAD_X,
                        pillY + (float) SheetSpec.LABEL_PAD_Y - ascent, text);
            }

            ByteArrayOutputStream out = new ByteArrayOutputStream();
            sheet.compress(Bitmap.CompressFormat.JPEG, SheetSpec.JPEG_QUALITY, out);
            // One retry, not a search: a second encode is another full pass over
            // eight megapixels and the frames are best effort anyway.
            if (out.size() > SheetSpec.JPEG_MAX_BYTES) {
                out = new ByteArrayOutputStream();
                sheet.compress(Bitmap.CompressFormat.JPEG, SheetSpec.JPEG_FALLBACK_QUALITY, out);
            }
            if (out.size() == 0 || out.size() > SheetSpec.JPEG_MAX_BYTES)
                throw new IllegalStateException("sheet exceeds upload limit");
            return out.toByteArray();
        } finally {
            sheet.recycle();
        }
    }

    /**
     * API 27 scales inside the decoder, which is the whole performance story: a
     * 1080x1920 frame is never materialised. Below that the frame arrives full
     * size and is scaled and released immediately, so peak memory is one frame
     * rather than twenty-five.
     */
    private static Bitmap frameAt(MediaMetadataRetriever retriever, long timeUs, int w, int h) {
        if (Build.VERSION.SDK_INT >= 27) {
            return retriever.getScaledFrameAtTime(timeUs, MediaMetadataRetriever.OPTION_CLOSEST_SYNC, w, h);
        }
        Bitmap full = retriever.getFrameAtTime(timeUs, MediaMetadataRetriever.OPTION_CLOSEST_SYNC);
        if (full == null) return null;
        float scale = Math.min(w / (float) full.getWidth(), h / (float) full.getHeight());
        Bitmap scaled = Bitmap.createScaledBitmap(full,
                Math.max(1, Math.round(full.getWidth() * scale)),
                Math.max(1, Math.round(full.getHeight() * scale)), true);
        if (scaled != full) full.recycle();
        return scaled;
    }

    private static double number(MediaMetadataRetriever retriever, int key) {
        try {
            String raw = retriever.extractMetadata(key);
            return raw == null ? 0 : Double.parseDouble(raw.trim());
        } catch (Exception e) {
            return 0;
        }
    }
}

File: ios/App/Shared/ContactSheet.swift
import AVFoundation
import UIKit

/**
 * The phone cuts the frames.
 *
 * Up to three JPEG grids of timestamped stills are what the visual reader gets
 * instead of a video, and cutting them here costs nothing: AVAssetImageGenerator
 * on a local file runs at roughly 12-64 ms a frame, so thirty-six frames are
 * about a second of the phone's own time and zero cents of anybody's API budget.
 *
 * Two constraints shape every decision below, and both come from the Share
 * Extension rather than the app: an app extension is killed somewhere around 120
 * MB, and it has no time to spare. So `maximumSize` makes AVFoundation decode
 * straight down to cell size — a full 1080x1920 frame is never held — and a sheet
 * is drawn and encoded the moment its twelve cells are full, which means peak
 * memory is twelve frames and one canvas (about 12 MB) rather than all
 * thirty-six. Interrupted sampling falls back to the server video route; a finished
 * prefix must never be represented as an overview of the complete video.
 *
 * Shared verbatim by the extension and by the in-app plugin; `SheetSpec` holds
 * every number and is mirrored on Android.
 */

struct SheetPage {
    let jpeg: Data
    let times: [Double]
    let cols: Int
    let rows: Int
}

struct ContactSheetResult {
    let pages: [SheetPage]
    let cellW: Int
    let cellH: Int
    /// Frames asked for, so a caller can log how much of the clip it actually saw.
    let requested: Int
    let duration: Double
}

enum ContactSheetError: Error {
    case noVideoTrack
    case noFrames
    case encodeFailed
}

enum ContactSheetBuilder {
    /// Below this, a sheet says less than nothing — better to save without frames
    /// and let the server decide whether to spend a Gemini read on the video.
    static let minUsableFrames = 4

    static func build(mp4: URL, duration hinted: Double, deadline: Date) async throws -> ContactSheetResult {
        let asset = AVURLAsset(url: mp4, options: [AVURLAssetPreferPreciseDurationAndTimingKey: false])
        let (display, measured) = try await describe(asset)
        // Decoder-measured duration is authoritative; page metadata can be stale.
        let duration = measured.isFinite && measured > 0 ? measured : hinted
        let cell = SheetSpec.cell(videoWidth: display.width, videoHeight: display.height)
        let wanted = SheetSpec.times(duration: duration)

        let generator = AVAssetImageGenerator(asset: asset)
        generator.appliesPreferredTrackTransform = true
        generator.maximumSize = CGSize(width: cell.w, height: cell.h)
        // Half a second of slack turns "decode from the last keyframe forward" into
        // "hand me the nearest keyframe", which is the whole difference between 12
        // ms and 200 ms a frame. What the model is told is the time it actually got.
        let slack = CMTime(seconds: SheetSpec.edgeInset, preferredTimescale: 600)
        generator.requestedTimeToleranceBefore = slack
        generator.requestedTimeToleranceAfter = slack

        var pages: [SheetPage] = []
        var chunk: [(time: Double, image: CGImage)] = []
        var last = -Double.greatestFiniteMagnitude
        var kept = 0
        var decoded = 0

        // Twelve cells make a sheet; the sheet is drawn and the frames released
        // before the next twelve are asked for.
        func flush() throws {
            guard !chunk.isEmpty, pages.count < SheetSpec.maxSheets else { chunk = []; return }
            pages.append(try render(frames: chunk, cell: cell))
            chunk = []
        }
        func take(_ time: Double, _ image: CGImage) throws -> Bool {
            guard time.isFinite, time >= 0, time <= duration else { return true }
            decoded += 1
            guard SheetSpec.accepts(time, after: last) else { return true }
            chunk.append((time, image)); last = time; kept += 1
            if chunk.count == SheetSpec.cellsPerSheet { try flush() }
            return pages.count < SheetSpec.maxSheets
        }

        let times = wanted.map { CMTime(seconds: $0, preferredTimescale: 600) }
        for await result in generator.images(for: times) {
            if case let .success(_, image, actual) = result {
                if try !take(actual.seconds, image) { break }
            }
            // The budget is the user's patience, not a correctness property:
            // whatever has arrived by now is the sheet, and the rest is dropped.
            if Date() >= deadline { break }
        }
        generator.cancelAllCGImageGeneration()
        try flush()

        // Never publish a timed-out prefix as an overview of the whole clip.
        guard decoded == wanted.count, kept >= minUsableFrames, !pages.isEmpty else { throw ContactSheetError.noFrames }
        return ContactSheetResult(pages: pages, cellW: cell.w, cellH: cell.h,
                                  requested: wanted.count, duration: duration)
    }

    // MARK: - drawing

    private static func render(frames: [(time: Double, image: CGImage)],
                               cell: (w: Int, h: Int)) throws -> SheetPage {
        let canvas = SheetSpec.canvas(count: frames.count, cell: cell)
        let format = UIGraphicsImageRendererFormat.default()
        // scale 1 because these are pixels for a model, not points for a screen;
        // at scale 3 the same sheet would be nine times the memory and no more
        // information. Opaque and standard-range for the same reason: no alpha
        // channel and no wide-gamut buffer to carry through the encode.
        format.scale = 1
        format.opaque = true
        format.preferredRange = .standard

        let size = CGSize(width: canvas.w, height: canvas.h)
        let renderer = UIGraphicsImageRenderer(size: size, format: format)
        let attributes: [NSAttributedString.Key: Any] = [
            .font: UIFont.systemFont(ofSize: SheetSpec.labelFontSize, weight: .bold),
            .foregroundColor: UIColor.white
        ]

        let draw: (UIGraphicsImageRendererContext) -> Void = { context in
            // Black under everything: the gutters, the bars beside an aspect-fitted
            // frame, and any cell the last row does not reach.
            UIColor.black.setFill()
            context.fill(CGRect(origin: .zero, size: size))

            for (index, frame) in frames.enumerated() {
                let slot = SheetSpec.frame(index: index, cell: cell)
                let box = CGRect(x: slot.x, y: slot.y, width: slot.w, height: slot.h)
                let w = CGFloat(frame.image.width), h = CGFloat(frame.image.height)
                if w > 0, h > 0 {
                    // Fit, never fill. A crop is exactly where the second kettlebell
                    // at the edge of the frame would have gone missing.
                    let scale = min(box.width / w, box.height / h)
                    let fitted = CGRect(x: box.midX - w * scale / 2, y: box.midY - h * scale / 2,
                                        width: w * scale, height: h * scale)
                    UIImage(cgImage: frame.image).draw(in: fitted)
                }

                let text = SheetSpec.label(seconds: frame.time) as NSString
                let measured = text.size(withAttributes: attributes)
                let pill = CGRect(
                    x: box.minX + SheetSpec.labelInset,
                    y: box.maxY - SheetSpec.labelInset - measured.height - SheetSpec.labelPadY * 2,
                    width: measured.width + SheetSpec.labelPadX * 2,
                    height: measured.height + SheetSpec.labelPadY * 2)
                UIColor(white: 0, alpha: SheetSpec.labelPillAlpha).setFill()
                UIBezierPath(roundedRect: pill, cornerRadius: SheetSpec.labelPillRadius).fill()
                text.draw(at: CGPoint(x: pill.minX + SheetSpec.labelPadX, y: pill.minY + SheetSpec.labelPadY),
                          withAttributes: attributes)
            }
        }

        var data = renderer.jpegData(withCompressionQuality: SheetSpec.jpegQuality, actions: draw)
        // One retry, not a search: a second encode is another full pass over a
        // megapixel and a half, and the frames are best-effort anyway.
        if data.count > SheetSpec.jpegMaxBytes {
            data = renderer.jpegData(withCompressionQuality: SheetSpec.jpegFallbackQuality, actions: draw)
        }
        guard !data.isEmpty, data.count <= SheetSpec.jpegMaxBytes else { throw ContactSheetError.encodeFailed }
        return SheetPage(jpeg: data, times: frames.map(\.time),
                         cols: min(SheetSpec.cols, frames.count),
                         rows: SheetSpec.rows(count: frames.count))
    }

    // MARK: - asset facts

    /// Display size after the track's rotation, and the duration, on both OS eras.
    private static func describe(_ asset: AVURLAsset) async throws -> (CGSize, Double) {
        guard let track = try await asset.loadTracks(withMediaType: .video).first else {
            throw ContactSheetError.noVideoTrack
        }
        let (natural, transform) = try await track.load(.naturalSize, .preferredTransform)
        let duration = try await asset.load(.duration)
        return (natural.applying(transform).absolute, duration.seconds)
    }

}

private extension CGSize {
    /// A 90-degree preferredTransform flips a size negative; only magnitude matters.
    var absolute: CGSize { CGSize(width: abs(width), height: abs(height)) }
}
